ops/cli.py

```diff
@@ -1,351 +1,343 @@
+
+
+class CLI:
+    def __init__(self):
+        """Set up the parser."""
+        parser = argparse.ArgumentParser(description="Deploy or connect to WFL infrastructure instances",
+                                         usage="%(prog)s [-h] [-d | -f] COMMAND [-e ENV] [-i INSTANCE] [...]")
+        parser.add_argument("-d", "--dry-run", action="store_true",
+                            help="Prevent COMMAND from enacting changes")
+        parser.add_argument("-f", "--force", action="store_true",
+                            help="Force COMMAND to run even if validation fails")
+        commands = parser.add_argument_group("COMMAND")
+        commands.add_argument("command", choices=["deploy", "connect", "info"],
+                              help="Deploy the local build to the instance, "
+                                   "connect to the instance's Cloud SQL, "
+                                   "or display instance config")
+        typical = parser.add_argument_group("typical arguments",
+                                            "(usually necessary to target a particular instance)")
+        typical.add_argument("-e", "--environment", default="gotc-dev", metavar="ENV",
+                             choices=["gotc-dev", "gotc-prod", "aou"],
+                             help="Specify 'gotc-deploy/deploy/{ENV}' with one of: %(choices)s")
+        typical.add_argument("-i", "--instance", default="dev",
+                             help="Provide the ID of the target instance in the environment")
+        specific = parser.add_argument_group("specific arguments",
+                                             "(rarely necessary, inferred from typical arguments)")
+        specific.add_argument("-v", "--version",
+                              help="Specify the exact version to use instead of the 'version' file")
+        specific.add_argument("--project",
+                              help="Specify the exact GCP project name instead of 'broad-{ENV}'")
+        specific.add_argument("--cloud-sql",
+                              help="Specify the exact Cloud SQL name instead of inferring from labels")
+        specific.add_argument("--zone",  # don't set default here so Config knows if this was set
+                              help="Specify the exact GCP cluster zone instead of inferring from labels")
+        specific.add_argument("--cluster",
+                              help="Specify the exact GKE cluster name instead of inferring from labels")
+        specific.add_argument("--namespace",
+                              help="Specify the exact K8s namespace instead of '{INSTANCE}-wfl'")
+        self.parser = parser
+
+    def execute(self) -> int:
+        """Run command."""
+        args = self.parser.parse_args()
+        for cmd in ["docker", "gcloud", "git", "helm", "java", "kubectl", "jq"]:
+            if not shutil.which(cmd):
+                error(f"{cmd} is not in PATH. Please install it!")
+        commands = {
```

Comment:
Why can't you manage this mapping in the parser? You've got two lists of commands to maintain now.

Reply:
For clarity, I'm taking about the `commands` mapping.

Reply:
`choices` doesn't operate on a mapping, you can give it a list of type T and a function from String to T. The closest I was able to get is this:
```
commands.add_argument("command",
                              type=lambda s: getattr(sys.modules[__name__], f"{s.title()}Command"),
                              choices=[DeployCommand, ConnectCommand, InfoCommand],
                              metavar="{deploy, connect, info}",
                              help="Deploy the local build to the instance, "
                                   "connect to the instance's Cloud SQL, "
                                   "or display instance config")
```
This does make the code you're referencing here cleaner: `config = args.command(args)`

Reply:
Alternatively, you could provid this function a `Dictionary<String, Callable>`, you could use that same dictionary to populate this parser as for dispatch. Then there's only one list to maintain.

Reply:
Fair enough. Abstracted out the mapping to the CLI class.